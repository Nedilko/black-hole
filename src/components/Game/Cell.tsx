import React from 'react';

type PropsType = {
  index: number;
  // onClick: (index: number) => void;
  isHole: boolean;
  holesNearCount: number;
  isOpen: boolean;
  onOpen: (index: number, isMark: boolean) => void;
};

const Cell = ({ index, isHole, holesNearCount, isOpen, onOpen }: PropsType) => {
  // const [isOpenCell, setIsOpenCell] = useState(isOpen);
  // const isRounded = isMine && isOpenCell;

  // const handleClick = useCallback(
  //   (e: any) => {
  //     e.preventDefault();
  //     console.log(`Cell ${index} clicked`);
  //     if (e.type === 'click') {
  //       setIsOpenCell(true);
  //     } else if (e.type === 'contextmenu') {
  //     }
  //   },
  //   [index]
  // );

  const handleClick = (e: any) => {
    e.preventDefault();
    onOpen(index, e.type === 'contextmenu');
  };

  const cellValue = isOpen ? (isHole ? 'ðŸ’¥' : holesNearCount) : '';

  return (
    <button
      onClick={handleClick}
      onContextMenu={handleClick}
      className={`w-12 h-12 ${isHole ? 'border hover:border-red-300' : ''} ${
        isOpen
          ? holesNearCount === 0
            ? 'rounded-md bg-gray-100'
            : 'rounded-full border-2'
          : 'bg-gray-200 hover:bg-gray-300 rounded-md'
      }`}
    >
      {cellValue}
    </button>
  );
};

export default React.memo(Cell);
