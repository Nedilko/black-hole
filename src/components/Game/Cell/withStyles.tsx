import { FC } from 'react';
import type { CellPropsType, WithStylesProps } from './index';

export type CellType = 'regular' | 'empty' | 'hole';

const CELL_STYLES = {
  HOLE: 'rounded-full border-2 border-red-800 bg-black cursor-default',
  EMPTY: 'border-2 rounded-md border-gray-700 bg-gray-700/10 cursor-default',
  REGULAR: 'rounded-full border-2 border-gray-300 cursor-default',
};

const withStyles = <T extends CellPropsType = CellPropsType>(
  WrappedCell: FC<T>
) => {
  const StyledCell = (props: Omit<T, keyof WithStylesProps>) => {
    const { holesNearCount, isHole, isOpen, isMarked } = props;

    let cellStyle = isHole
      ? CELL_STYLES.HOLE
      : holesNearCount === 0
      ? CELL_STYLES.EMPTY
      : CELL_STYLES.REGULAR;

    const markedStyle = isMarked ? 'border-cyan-600/90' : 'border-gray-400';
    if (!isOpen) {
      cellStyle = `rounded-md border-2 ${markedStyle} bg-gradient-radial hover:bg-blue-900/30 hover:from-blue-900/30 from-black/70 via-black/10 to-black/10"`;
    }
    return <WrappedCell {...(props as T)} cellStyle={cellStyle} />;
  };

  StyledCell.displayName = `StyledCell`;
  return StyledCell;
};

export default withStyles;
