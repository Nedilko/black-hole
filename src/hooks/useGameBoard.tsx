import { useMemo, useState } from 'react';
import { GameBoard, IBoardWithCells, IBoardSize } from '../game/field';

export const useGameBoard = (
  size: IBoardSize,
  holesCount: number,
  onFinish: () => void,
  onOpenCell: () => void
): [IBoardWithCells, number] => {
  const board: IBoardWithCells = useMemo(
    () =>
      GameBoard.create(
        size,
        holesCount,
        () => {
          setOpenedIndexes(board.openedCellIndexes.length);
          onOpenCell();
        },
        onFinish
      ),
    [holesCount, size]
  );

  const [openedIndexesCount, setOpenedIndexes] = useState(
    board.openedCellIndexes.length
  );

  return [board, openedIndexesCount];
};
