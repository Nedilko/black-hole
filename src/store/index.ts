import {
  configureStore,
  combineReducers,
  PreloadedState,
} from '@reduxjs/toolkit';
import gameReducer from './gameSlice';
import fieldReducer from './fieldSlice';

const rootReducer = combineReducers({
  game: gameReducer,
  field: fieldReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
