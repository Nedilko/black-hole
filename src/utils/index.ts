import { getTime } from './time';
import {
  readFromLocalstorage,
  writeToLocalstorage,
} from './localstorageAdapter';
import { getDefaultResults } from './game';
import { loadResults, writeResults } from './dataAdapter';

export {
  getTime,
  readFromLocalstorage,
  writeToLocalstorage,
  getDefaultResults,
  loadResults,
  writeResults,
};
