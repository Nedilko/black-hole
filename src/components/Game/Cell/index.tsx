import React, { MouseEventHandler } from 'react';
import withInterctivity from './withIntercativity';
import withStyles from './withStyles';

export interface WithInteractivityProps {
  handleOpen: (index: number) => void;
  handleMarkCell: (index: number) => void;
}

export interface WithStylesProps {
  cellStyle?: string;
}

export interface CellPropsType extends WithInteractivityProps, WithStylesProps {
  isOpen: boolean;
  isHole: boolean;
  isMarked: boolean;
  holesNearCount: number;
  index: number;
}

const Cell = ({
  index,
  handleOpen,
  handleMarkCell,
  holesNearCount,
  isHole,
  isOpen,
  cellStyle = '',
}: CellPropsType) => {
  const handleLeftClick: MouseEventHandler<HTMLButtonElement> = () => {
    handleOpen(index);
  };

  const handleRightClick: MouseEventHandler<HTMLButtonElement> = (event) => {
    event.preventDefault();
    handleMarkCell(index);
  };

  const canShowValue = holesNearCount !== 0 && !isHole;
  const displayValue = isOpen ? holesNearCount : '';

  return (
    <button
      onClick={handleLeftClick}
      onContextMenu={handleRightClick}
      className={`text-gray-300 text-2xl font-medium w-10 h-10 opacity-90 transition ease-in-out duration-150 ${cellStyle}`}
    >
      {canShowValue && displayValue}
    </button>
  );
};

export default React.memo(withInterctivity(withStyles(Cell)));
