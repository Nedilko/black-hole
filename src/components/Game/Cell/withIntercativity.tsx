import { FC, useCallback } from 'react';
import { useAppDispatch } from '../../../hooks';
import { toggleMarkCell, openCell } from '../../../store/fieldSlice';
import type { WithInteractivityProps } from './index';

const withInterctivity = <
  T extends WithInteractivityProps = WithInteractivityProps
>(
  WrappedComponent: FC<T>
) => {
  const WithInteractivity = (props: Omit<T, keyof WithInteractivityProps>) => {
    const dispatch = useAppDispatch();
    const handleOpenCell = useCallback(
      (cellIndex: number) => {
        dispatch(openCell(cellIndex));
      },
      [dispatch]
    );

    const handleMarkCell = useCallback(
      (cellIndex: number) => {
        dispatch(toggleMarkCell(cellIndex));
      },
      [dispatch]
    );

    return (
      <WrappedComponent
        {...(props as T)}
        handleOpen={handleOpenCell}
        handleMarkCell={handleMarkCell}
      />
    );
  };

  WithInteractivity.displayName = `WithInteractivity`;

  return WithInteractivity;
};

export default withInterctivity;
