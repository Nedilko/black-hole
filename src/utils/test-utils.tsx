import { RenderOptions } from '@testing-library/react';
import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { PreloadedState } from 'redux';
import { AppStore, RootState, setupStore } from '../store';
import { render } from '@testing-library/react';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = setupStore(initialState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
};
