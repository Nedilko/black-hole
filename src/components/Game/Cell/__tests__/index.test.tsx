import Cell from '..';
import { renderWithProviders } from '../../../../utils/test-utils';
import * as hooks from '../../../../hooks';
import * as actions from '../../../../store/fieldSlice';
import userEvent from '@testing-library/user-event';
import { fireEvent, screen } from '@testing-library/react';

describe('Cell', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render Cell', () => {
    const { container } = renderWithProviders(
      <Cell
        index={0}
        isHole={false}
        isOpen={false}
        isMarked={false}
        holesNearCount={3}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render hole Cell', () => {
    const { container } = renderWithProviders(
      <Cell
        index={0}
        isHole={true}
        isOpen={false}
        isMarked={false}
        holesNearCount={3}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render open Cell', () => {
    const { container } = renderWithProviders(
      <Cell
        index={0}
        isHole={false}
        isOpen={true}
        isMarked={false}
        holesNearCount={3}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render marked Cell', () => {
    const { container } = renderWithProviders(
      <Cell
        index={0}
        isHole={false}
        isOpen={false}
        isMarked={true}
        holesNearCount={3}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render not empty Cell', () => {
    const { container } = renderWithProviders(
      <Cell
        index={0}
        isHole={false}
        isOpen={true}
        isMarked={false}
        holesNearCount={0}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should handle cell left click', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    const openCellAction = jest.spyOn(actions, 'openCell');

    renderWithProviders(
      <Cell
        index={2}
        isHole={false}
        isOpen={false}
        isMarked={false}
        holesNearCount={0}
      />
    );

    userEvent.click(screen.getByRole('button'));
    expect(openCellAction).toHaveBeenCalledTimes(1);
    expect(openCellAction).toHaveBeenCalledWith(2);
  });

  it('should handle cell right click', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    const toggleMarkCellAction = jest.spyOn(actions, 'toggleMarkCell');

    renderWithProviders(
      <Cell
        index={2}
        isHole={false}
        isOpen={false}
        isMarked={false}
        holesNearCount={0}
      />
    );

    fireEvent.contextMenu(screen.getByRole('button'));
    expect(toggleMarkCellAction).toHaveBeenCalledTimes(1);
    expect(toggleMarkCellAction).toHaveBeenCalledWith(2);
  });
});
