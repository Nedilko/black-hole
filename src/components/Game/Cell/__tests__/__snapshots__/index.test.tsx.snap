// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render Cell 1`] = `
<div>
  <button
    class="text-gray-300 text-2xl font-medium w-10 h-10 opacity-90 transition ease-in-out duration-150 rounded-md border-2 border-gray-400 bg-gradient-radial hover:bg-blue-900/30 hover:from-blue-900/30 from-black/70 via-black/10 to-black/10\\""
  />
</div>
`;

exports[`Cell should render not empty Cell 1`] = `
<div>
  <button
    class="text-gray-300 text-2xl font-medium w-10 h-10 opacity-90 transition ease-in-out duration-150 border-2 rounded-md border-gray-700 bg-gray-700/10 cursor-default"
  />
</div>
`;
